SPECS/dracut/add-livenet-download-service.patch

```diff
@@ -0,0 +1,56 @@
+From 3055f6f900c5718d93bb68a4cdf0798ac0f55c14 Mon Sep 17 00:00:00 2001
+From: George Mileka <gmileka@microsoft.com>
+Date: Fri, 30 Aug 2024 18:17:00 -0700
+Subject: [PATCH] Enable netroot for systemd-networkd - temporary fix till
+ upgrading to 103+.
+
+Dracut version 102 does not implement the code path that handles livenet
+rootfs download (i.e. by calling /sbin/livenetroot) when systemd-networkd
+is the underlying networking manager.
+
+This has been implemented in 103 (see https://github.com/dracut-ng/dracut-ng/pull/388)
+
+As a mitigation for 102, this patch allows Dracut's livenet module to schedule
+a daemon to run after the network stack is up and running. The daemon will
+then invoke the usual livenet rootfs handling code to download and pivot to
+the new rootfs.
+---
+ modules.d/90livenet/module-setup.sh  | 2 ++
+ modules.d/90livenet/parse-livenet.sh | 8 ++++++++
+ 2 files changed, 10 insertions(+)
+
+diff --git a/modules.d/90livenet/module-setup.sh b/modules.d/90livenet/module-setup.sh
+index db0def50..74bacd6c 100755
+--- a/modules.d/90livenet/module-setup.sh
++++ b/modules.d/90livenet/module-setup.sh
+@@ -17,6 +17,8 @@ install() {
+     inst_hook cmdline 29 "$moddir/parse-livenet.sh"
+     inst_hook initqueue/online 95 "$moddir/fetch-liveupdate.sh"
+     inst_script "$moddir/livenetroot.sh" "/sbin/livenetroot"
++    inst_simple "$moddir/liveos-artifacts-download.service" "/etc/systemd/system/liveos-artifacts-download.service"
++    inst_script "$moddir/liveos-artifacts-download.sh" "/sbin/liveos-artifacts-download"
+     if dracut_module_included "systemd-initrd"; then
+         inst_script "$moddir/livenet-generator.sh" "$systemdutildir"/system-generators/dracut-livenet-generator
+     fi
+diff --git a/modules.d/90livenet/parse-livenet.sh b/modules.d/90livenet/parse-livenet.sh
+index a1d14a8f..996f5210 100755
+--- a/modules.d/90livenet/parse-livenet.sh
++++ b/modules.d/90livenet/parse-livenet.sh
+@@ -27,6 +27,14 @@ if get_url_handler "$liveurl" > /dev/null; then
+     root="livenet" # quiet complaints from init
+     # shellcheck disable=SC2034
+     rootok=1
++
++    enableAzureLinuxDownloader=$(getarg rd.live.azldownloader=)
++
++    if [[ "$enableAzureLinuxDownloader" == "enable" ]]; then
++        systemctl enable liveos-artifacts-download
++        systemctl start liveos-artifacts-download &
```

Comment:
You typically shouldn't manually start systemd services. Instead, you should add a `WantedBy` to the `.service` file and then enable the service. Then the service will be scheduled automatically by systemd.

To enable a service, in the `module-setup.sh` file run `$SYSTEMCTL -q --root "$initdir" enable <name>.service`

For the `WantedBy`, you'll want to check the [dracut.bootup](https://man7.org/linux/man-pages/man7/dracut.bootup.7.html). But I think you may want `WantedBy=sysinit.target`.

In terms of `getarg rd.live.azldownloader=`, you can probably check that in `liveos-artifacts-download.sh` and return early if it isn't present.

Reply:
We actually don't want the service enabled all the time at initrd generation time. Instead, we want it enabled only when it's a live netroot setup - hence, the enablement in the parse file and not in the module-setup.sh.

The problem with always having it enabled is that the user could have a network manager other than systemd-networkd enabled, and then specifies `root=live:http:...`. This will result in triggering both handlers (livenet and ours) at the same time if our new daemon is always enabled.

For the condition, we can't move to the script because we need it at parse time to enable the daemon.

Reply:
Fair enough. But it might be good to add a comment.